### DISCUSSION TITLE
How to auto save item

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#11666] [Archive]
Hi,
With long forms, I'd like to send the data during the editing process.
For instance, every time an accordion is closed, or a control is left.

For that, I can make an extension that will listen to such events.
But what would be the code to just **Update** the data in the database _(doing the same as_ Save _but without closing the form)_?

_Just to mention, that the button Save already does this on_ `Project Settings` _: it updates but, on purpose, does not close the form_

**Comment by benhaynes** [comment]
Have you seen the additional options within the save context menu? If you click on the three dots to the right of the save button you'll get new options for things like "Save and Stay"... which can also be accessed with a keyboard shortcut, eg: `CMD+S`

**Reply by Pachat** [reply]
Yes. And that's good to have.
But what I need is to send "Save and Stay" programmatically on events like closing an accordion.

**Reply by benhaynes** [reply]
We'll have to see if we can surface those "soft" events in the system, then you can tie logic into those event hooks. @rijkvanzanten — would that even be possible?

**Reply by rijkvanzanten** [reply]
No, that's just a regular save at that point, so no different from hitting the save button manually 🤔

**Reply by Pachat** [reply]
Except that, in this case, we don't want the form to close: the user wants to continue to edit the form, 
(just as it is the case on "Project Settings" where the form don't close when clicking on the Save button, may be because it is a Unique)

Why do we want this feature? On very long forms, the user may loose the connection or be out of power, etc.
He may accept a short loss of work on a few inputs, but not on the whole work he/she has done from the beginning.

**Reply by benhaynes** [reply]
Thank you for clarifying. For now, the "Save and Stay" button and keyboard shortcut will have to do. In the future, we're planning to implement automatic saving stages changes to localstorage (just like GitHub) so they aren't accidentally lost. We can look into surfacing new event hooks for the logic you need, but would need to see more users agree this is useful to them first.

**Reply by Pachat** [reply]
Automatic saving stages changes to local storage will no doubt be an improvement.
But the case here is for long forms. You may think at the applying form to a U. 
You don't want to fill at once and may have to consider using different devices along the way to fulfill all the requirements. 
For instance, you may begin with you tablet, then continue later with your laptop where you have files to upload, etc. 
Then make some updates on your phone.

The application I'm trying to port to Directus has very long forms and behaves that way:

| Local action   | Distant action |
|---|---|
| Click on the `Create` button: by default, the form's status is `Draft`<br>_( on later sessions, once the form created, click to update the form )_ |  no action<br>_( retrieve the form )_ | 
| First input &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `onblur` =><br>&nbsp; | Creates in the database a row with the input.<br><= The new `id` is sent back to the form | 
| Subsequent input &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;`onblur`  =><br> | updates the corresponding field of the `id` row |
| _repeat previous line until the form is ready_ |   |
| Once the form is ready, set the form's status to `Published`<br/>and click the `Save `button &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; IF the requirements (*) are OK:  | <br />updates the form as `Published  `|

(*) _The requirements such as_ `Required` _were not blocking while the form was in_ `Draft` _status_

This leads to be able to send the input value on the `onblur` event of every input

**Reply by VaZark** [reply]
Hi, I'm stepping in to give my 2 cents as someone who worked on forms with 40+ fields.

What we figured out early was that just looking at very long forms discouraged users and hurt our conversion rates. So we split the fields into multiple pages to create a multi-step form. This coincidentally solved the problem discussed here since we updated the data in the db/collecton for each "step". So the user wasn't too surprised when data was missing on the current step they were working on.

I'd suggest you can try creating a generic module for generating forms for collections with the ability to be split as multi-step forms (via field groups or manually assigned) and save the data after each page. It also gives you more power on how to handle onblur and onsubmit events as well as customise it later down the road.

**Reply by Pachat** [reply]
@VaZark  Yes. That's a good idea.
Is it possible to create multi-step forms with Directus itself?
For instance, the Save button currently brings back to the collection's list. How make it going to the next step?
And how to traverse the steps forth and back?

**Reply by VaZark** [reply]
AFAIK, no. You'll have to create a module extension to create the form yourself. It'll be classic Vue from that point. 

For the page traversal, create a parent component that keeps track of the form page and renders it. Then create different components for each page with the corresponding fields. Then when user clicks `Save` or `Next`. Save the collection, then update the Parent to render the next component.

Or use a router.

**Comment by Pachat** [comment]
@rijkvanzanten Thank you for your suggestion.  What Directus method could I include in my event handler to send a "Save and Stay"?

**Reply by rijkvanzanten** [reply]
There's no event emitter on the client side, so there's no way to trigger a Save and Stay operation from an interface

**Reply by Pachat** [reply]
Isn't there is an event we can create by JavaScript on clicking the accordion? 
So when we have only one group open, the user has to click to navigate from one group to the other.
That is the idea. May be I am wrong.

**Reply by rijkvanzanten** [reply]
There's no global event emitter / listener system in the app currently, so you can't just attach any operation to any event on a component 🤔 

(Custom) interfaces currently don't have access to the parents save operations either, so I don't think that could work in the current setup

**Reply by Pachat** [reply]
@rijkvanzanten Understood. Thanks for the clarification.
